File: tools/cli/internal/capability/capability_test.go
package capability

import (
	"encoding/json"
	"reflect"
	"testing"
)

// declare builds a valid declaration accepting the given names.
func declare(t *testing.T, accepts ...string) *Declaration {
	t.Helper()
	m := map[string]json.RawMessage{}
	for _, n := range accepts {
		m[n] = json.RawMessage(`{}`)
	}
	d := &Declaration{Schema: "1.0.0", Surface: Surface{Kind: "panel", Instance: "dev-1"}, Accepts: m}
	if err := d.Validate(); err != nil {
		t.Fatalf("fixture declaration invalid: %v", err)
	}
	return d
}

func TestClassify(t *testing.T) {
	cases := map[string]Class{
		"connected":    ClassLifecycle,
		"navigate":     ClassPresentationCommand,
		"reload":       ClassPresentationCommand,
		"device_cmd":   ClassPresentationCommand,
		"input_action": ClassPresentationCommand,
		"draft":        ClassPresentationCommand,
		"message":      ClassStateReport,
		"history":      ClassStateReport,
		"presence_map": ClassStateReport,
		"tool_event":   ClassStateReport,
		// Unknown names are state reports by construction: a new event is
		// deliverable-to-everyone until deliberately admitted to the gated
		// class.
		"some_future_event": ClassStateReport,
	}
	for event, want := range cases {
		if got := Classify(event); got != want {
			t.Errorf("Classify(%q) = %v, want %v", event, got, want)
		}
	}
}

func TestPresentationCommandsRosterIsTheFiveQ2dNames(t *testing.T) {
	want := []string{"device_cmd", "draft", "input_action", "navigate", "reload"}
	if got := PresentationCommands(); !reflect.DeepEqual(got, want) {
		t.Fatalf("PresentationCommands() = %v, want %v", got, want)
	}
}

// The gate table from docs/interface-capabilities.md, row by row.
func TestDecide(t *testing.T) {
	declared := declare(t, "navigate", "draft")
	cases := []struct {
		name    string
		decl    *Declaration
		event   string
		verdict Verdict
		reason  Reason
	}{
		{"legacy client gets commands", nil, "reload", VerdictDeliver, ReasonLegacy},
		{"legacy client gets reports", nil, "message", VerdictDeliver, ReasonLegacy},
		{"declared + accepted command", declared, "navigate", VerdictDeliver, ReasonAccepted},
		{"declared + unaccepted command suppresses", declared, "reload", VerdictSuppress, ReasonNotAccepted},
		{"declared client still gets reports", declared, "message", VerdictDeliver, ReasonUngated},
		{"declared client still gets lifecycle", declared, "connected", VerdictDeliver, ReasonUngated},
		{"unknown event delivers to declared client", declared, "some_future_event", VerdictDeliver, ReasonUngated},
	}
	for _, c := range cases {
		got := Decide(c.decl, c.event)
		if got.Verdict != c.verdict || got.Reason != c.reason {
			t.Errorf("%s: Decide → (%v, %v), want (%v, %v)", c.name, got.Verdict, got.Reason, c.verdict, c.reason)
		}
		if got.Event != c.event {
			t.Errorf("%s: decision echoes event %q, want %q", c.name, got.Event, c.event)
		}
	}
}

// The subtraction invariant: across every event name in every class, a
// declaration never causes a delivery the legacy client would not get —
// it can only remove them.
func TestDeclarationOnlySubtracts(t *testing.T) {
	declared := declare(t, "navigate")
	events := append(PresentationCommands(), "connected", "message", "history", "unheard_of")
	for _, ev := range events {
		legacy := Decide(nil, ev)
		if legacy.Verdict != VerdictDeliver {
			t.Fatalf("legacy client suppressed on %q — grandfathering broken", ev)
		}
		got := Decide(declared, ev)
		if got.Verdict == VerdictDeliver && legacy.Verdict != VerdictDeliver {
			t.Fatalf("declaration added a delivery on %q", ev)
		}
	}
}

func TestRecognize(t *testing.T) {
	d := declare(t, "navigate", "draft", "teleport", "hologram")
	recognized, unknown := d.Recognize()
	if want := []string{"draft", "navigate"}; !reflect.DeepEqual(recognized, want) {
		t.Errorf("recognized = %v, want %v", recognized, want)
	}
	if want := []string{"hologram", "teleport"}; !reflect.DeepEqual(unknown, want) {
		t.Errorf("unknown = %v, want %v", unknown, want)
	}

	// Empty accepts: both splits present and empty, never nil — the
	// connected-echo must serialize as [] not null.
	recognized, unknown = declare(t).Recognize()
	if recognized == nil || unknown == nil || len(recognized)+len(unknown) != 0 {
		t.Errorf("empty declaration Recognize = (%v, %v), want ([], [])", recognized, unknown)
	}
}

File: packages/server/src/capability.test.ts
import { describe, expect, test } from "bun:test"
import {
  MAX_ACCEPTS, MAX_DECLARATION_BYTES, MAX_TOKENS, PRESENTATION_COMMANDS,
  countSuppressed, parseDeclaration, recognize, shouldDeliver, suppressedCounts,
  type CapabilityDeclaration,
} from "./capability"

// Mirror of tools/cli/internal/capability's suite — the Go engine is the
// normative contract; these tests keep the TS mirror honest against it.

const valid = () => JSON.stringify({
  schema: "1.0.0",
  surface: { kind: "panel", instance: "dev-42" },
  accepts: { navigate: {}, reload: {} },
  content: ["text", "images"],
  interactions: ["select", "compose"],
})

function declare(...accepts: string[]): CapabilityDeclaration {
  const result = parseDeclaration(JSON.stringify({
    schema: "1.0.0",
    surface: { kind: "panel", instance: "dev-1" },
    accepts: Object.fromEntries(accepts.map(n => [n, {}])),
  }))
  if ("error" in result) throw new Error(`fixture declaration invalid: ${result.error}`)
  return result.decl
}

describe("parseDeclaration", () => {
  test("accepts a valid declaration", () => {
    const result = parseDeclaration(valid())
    if ("error" in result) throw new Error(result.error)
    expect(result.decl.surface).toEqual({ kind: "panel", instance: "dev-42" })
    expect(Object.keys(result.decl.accepts).length).toBe(2)
    expect(result.decl.content.length).toBe(2)
    expect(result.decl.interactions.length).toBe(2)
  })

  test("ignores unknown top-level fields (LSP posture)", () => {
    const raw = valid().replace('"schema"', '"from_the_future": {"x": 1}, "schema"')
    expect("decl" in parseDeclaration(raw)).toBe(true)
  })

  const rejections: Array<[string, string]> = [
    ["not json",                  '{"schema": '],
    ["missing schema",            '{"surface": {"kind": "panel"}}'],
    ["malformed schema",          '{"schema": "1.0", "surface": {"kind": "panel"}}'],
    ["prerelease schema",         '{"schema": "1.0.0-rc1", "surface": {"kind": "panel"}}'],
    ["unsupported major",         '{"schema": "2.0.0", "surface": {"kind": "panel"}}'],
    ["missing kind",              '{"schema": "1.0.0", "surface": {}}'],
    ["uppercase kind",            '{"schema": "1.0.0", "surface": {"kind": "Panel"}}'],
    ["bad instance",              '{"schema": "1.0.0", "surface": {"kind": "panel", "instance": "a b"}}'],
    ["accept name leading digit", '{"schema": "1.0.0", "surface": {"kind": "panel"}, "accepts": {"9lives": {}}}'],
    ["accept name uppercase",     '{"schema": "1.0.0", "surface": {"kind": "panel"}, "accepts": {"Navigate": {}}}'],
    ["accepts as array",          '{"schema": "1.0.0", "surface": {"kind": "panel"}, "accepts": ["navigate"]}'],
    ["bad content token",         '{"schema": "1.0.0", "surface": {"kind": "panel"}, "content": ["text/plain"]}'],
    ["bad interactions token",    '{"schema": "1.0.0", "surface": {"kind": "panel"}, "interactions": ["click!"]}'],
  ]
  for (const [name, raw] of rejections) {
    test(`rejects ${name}`, () => expect("error" in parseDeclaration(raw)).toBe(true))
  }

  test("rejects oversized declarations", () => {
    const padded = valid().replace('"dev-42"', `"dev-42", "pad": "${"x".repeat(MAX_DECLARATION_BYTES)}"`)
    expect("error" in parseDeclaration(padded)).toBe(true)
  })

  test("rejects over-cap entry counts", () => {
    const entries = Array.from({ length: MAX_ACCEPTS + 1 }, (_, i) => [`cap_${i}`, {}])
    const overAccepts = JSON.stringify({ schema: "1.0.0", surface: { kind: "panel" }, accepts: Object.fromEntries(entries) })
    expect("error" in parseDeclaration(overAccepts)).toBe(true)

    const tokens = Array.from({ length: MAX_TOKENS + 1 }, (_, i) => `tok_${i}`)
    const overTokens = JSON.stringify({ schema: "1.0.0", surface: { kind: "panel" }, content: tokens })
    expect("error" in parseDeclaration(overTokens)).toBe(true)
  })

  test("preserves accepts detail objects", () => {
    const result = parseDeclaration('{"schema": "1.0.0", "surface": {"kind": "panel"}, "accepts": {"device_cmd": {"cmds": ["flash"]}}}')
    if ("error" in result) throw new Error(result.error)
    expect(result.decl.accepts.device_cmd).toEqual({ cmds: ["flash"] })
  })
})

describe("shouldDeliver", () => {
  const declared = declare("navigate", "draft")

  // The gate table from docs/interface-capabilities.md, row by row.
  test("legacy client gets everything", () => {
    expect(shouldDeliver(undefined, "reload")).toBe(true)
    expect(shouldDeliver(undefined, "message")).toBe(true)
  })
  test("declared + accepted command delivers", () => expect(shouldDeliver(declared, "navigate")).toBe(true))
  test("declared + unaccepted command suppresses", () => expect(shouldDeliver(declared, "reload")).toBe(false))
  test("state reports and lifecycle stay ungated", () => {
    expect(shouldDeliver(declared, "message")).toBe(true)
    expect(shouldDeliver(declared, "connected")).toBe(true)
    expect(shouldDeliver(declared, "some_future_event")).toBe(true)
  })

  test("a declaration only subtracts deliveries", () => {
    const narrow = declare("navigate")
    for (const ev of [...PRESENTATION_COMMANDS, "connected", "message", "history", "unheard_of"]) {
      expect(shouldDeliver(undefined, ev)).toBe(true)
      if (shouldDeliver(narrow, ev)) expect(shouldDeliver(undefined, ev)).toBe(true)
    }
  })
})

describe("recognize", () => {
  test("splits accepts into recognized and unknown, sorted", () => {
    const { recognized, unknown } = recognize(declare("navigate", "draft", "teleport", "hologram"))
    expect(recognized).toEqual(["draft", "navigate"])
    expect(unknown).toEqual(["hologram", "teleport"])
  })
  test("empty accepts yields empty arrays, never null", () => {
    expect(recognize(declare())).toEqual({ recognized: [], unknown: [] })
  })
})

test("suppression counters accumulate and report sorted", () => {
  // Deltas, not absolutes: the counter map is module state shared with any
  // other test that exercises the gate.
  const before = suppressedCounts()
  countSuppressed("reload")
  countSuppressed("reload")
  countSuppressed("draft")
  const after = suppressedCounts()
  expect((after.reload ?? 0) - (before.reload ?? 0)).toBe(2)
  expect((after.draft ?? 0) - (before.draft ?? 0)).toBe(1)
  expect(Object.keys(after)).toEqual(Object.keys(after).toSorted())
})
